hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MembershipManager.java

```diff
@@ -0,0 +1,895 @@
+        } finally {
+            clusterServiceLock.unlock();
+        }
+    }
+
+    // called under cluster service lock
+    private void retainSuspectedMembers(MemberMap memberMap) {
+        Iterator<Address> it = suspectedMembers.iterator();
+        while (it.hasNext()) {
+            Address suspectedAddress = it.next();
+            if (!memberMap.contains(suspectedAddress)) {
+                logger.fine("Removing suspected address " + suspectedAddress + ", it's no longer a member.");
+                it.remove();
+            }
+        }
+    }
+
+    boolean isMemberSuspected(Address address) {
+        return suspectedMembers.contains(address);
+    }
+
+    boolean clearMemberSuspicion(Address address, String reason) {
+        clusterServiceLock.lock();
+        try {
+            if (!suspectedMembers.contains(address)) {
+                 return true;
+            }
+
+            MemberMap memberMap = getMemberMap();
+            Address masterAddress = clusterService.getMasterAddress();
+            if (memberMap.isBeforeThan(address, masterAddress)) {
+                logger.fine("Not removing suspicion of " + address + " since it is before than current master "
+                        + masterAddress + " in member list.");
+                return false;
+            }
+
+            boolean removed = suspectedMembers.remove(address);
+            if (removed && logger.isInfoEnabled()) {
+                logger.info("Removed suspicion from " + address + ". Reason: " + reason);
+            }
+
+        } finally {
+            clusterServiceLock.unlock();
+        }
+        return true;
+    }
+
+    void handleExplicitSuspicionTrigger(Address caller, int callerMemberListVersion,
+            MembersViewMetadata suspectedMembersViewMetadata) {
+        clusterServiceLock.lock();
+        try {
+            Address masterAddress = clusterService.getMasterAddress();
+            int memberListVersion = getMemberListVersion();
+
+            if (!(masterAddress.equals(caller) && memberListVersion == callerMemberListVersion)) {
+                logger.fine("Ignoring explicit suspicion trigger for " + suspectedMembersViewMetadata
+                        + ". Caller: " + caller + ", caller member list version: " + callerMemberListVersion
+                        + ", known master: " + masterAddress + ", local member list version: " + memberListVersion);
+                return;
+            }
+
+            clusterService.sendExplicitSuspicion(suspectedMembersViewMetadata);
+        } finally {
+            clusterServiceLock.unlock();
+        }
+    }
+
+    void handleExplicitSuspicion(MembersViewMetadata expectedMembersViewMetadata, Address suspectedAddress) {
+        clusterServiceLock.lock();
+        try {
+            MembersViewMetadata localMembersViewMetadata = createLocalMembersViewMetadata();
+            if (!localMembersViewMetadata.equals(expectedMembersViewMetadata)) {
```

Comment:
How come we ignore the request if the metadata isn't the same? Doesn't explicit suspicion occur when 2 members have different views - e.g. one member thinks it's master and sends a finalizeJoin operation or a member sends a heartbeat operation to the master but the sender is no longer a part of the cluster?